File: ugs/cloud-code/telemetry.js
const { DataApi } = require("@unity-services/cloud-save-1.4");

// ⚠️⚠️ THIS FILE IS `TelemetryRules.cs` WRITTEN A SECOND TIME, AND THE C# IS THE SPECIFICATION.
// The same trade `match-record.js` records about `ProfileRules` (`docs/TODO.md` § 89.6), applied
// to a smaller file. Cloud Code cannot import the C# and the C# cannot run here, so every
// constant below names the core member it mirrors and `CareerAndCloudCodeTests` fails if the
// numbers split. When the two disagree, THIS is the bug.
//
// ⚠️⚠️ AND THE EVENT NAMES ARE A CONTRACT, NOT A LIST. `FUTURE.md` § 19.3: *"a renamed event is a
// broken history"*. A name that changes does not produce an error anywhere; it produces a counter
// that starts at zero and a year of data that no longer joins to it. The server refuses a name it
// does not know so a typo in a new call site is a failure rather than a silent second series.
// `docs/TODO.md` § 90.3 is the contract in prose.

const KEY = "telemetry";
const ROLLUP_ID = "telemetry-rollup";
const ROLLUP_KEY = "totals";

// Mirrors `TelemetryRules.MaxEventsPerBatch` and `TelemetryRules.MaxParametersPerEvent`.
const MAX_EVENTS_PER_BATCH = 64;
const MAX_PARAMETERS_PER_EVENT = 8;
const MAX_PARAMETER_LENGTH = 32;

// Mirrors `TelemetryEvents.Funnel`, in order. ⚠️ APPEND ONLY, and the ORDER IS THE MEANING:
// "furthest step reached" is an index comparison, so inserting one in the middle rewrites what
// every stored profile is claiming. `FUTURE.md` § 0.5 rule 5 for wire-facing lists is this rule.
const FUNNEL = [
    "first_launch",
    "first_sign_in",
    "first_menu",
    "first_queue",
    "first_match_started",
    "first_match_finished",
];

// Mirrors `TelemetryEvents.All`. Anything not in here is refused.
const EVENTS = FUNNEL.concat([
    "session_start",
    "session_end",
    "match_started",
    "match_finished",
    "match_left",
    "pick",
    "settings_snapshot",
    "disconnect",
]);

function text(value, max) {
    const raw = String(value === undefined || value === null ? "" : value);
    let out = "";
    for (let i = 0; i < raw.length && out.length < max; i++) {
        const code = raw.charCodeAt(i);
        if (code < 32 || code === 127) continue;
        out += raw[i];
    }
    return out.trim();
}

// ⚠️⚠️ THE ONLY THING THAT MAY REACH THIS SCRIPT IS A NAME FROM THE LIST AND A HANDFUL OF SHORT
// LABELS. `FUTURE.md` § 19.3: *"No personally identifying field in any event, ever."* The client
// enforces the same rule in `TelemetryRules`, and this enforces it again because the client is
// the half somebody can edit. There is deliberately no free-text field: a parameter is a bucket
// label like `hero_strike` or `p95`, never a sentence somebody typed.
//
// ⚠️ AND THE PLAYER IS NEVER NAMED IN THE PAYLOAD AT ALL. The caller's id comes from
// `context.playerId`, which the client cannot set, so there is no identifier in the batch to
// leak, to get wrong, or to have to strip later.
// ⚠️⚠️ THIS FUNCTION WAS CALLED `parameters` AND THAT ONE NAME BROKE THE DEPLOY, SILENTLY.
// `docs/TODO.md` § 90.5: with a top-level `function parameters(...)` in the file, `ugs deploy`
// uploaded the code and then reported `params: []` for the whole script, so `action` was stripped
// from every call and every request landed on the default branch. Nothing failed, nothing warned,
// and the endpoint answered normally from the wrong branch. Bisected one function at a time
// against the live service: renaming it is the entire fix. **Do not name anything at the top level
// of a Cloud Code script `parameters`.**
function columnsFrom(raw) {
    const result = {};
    if (!raw || typeof raw !== "object") return result;

    let kept = 0;
    for (const key of Object.keys(raw)) {
        if (kept >= MAX_PARAMETERS_PER_EVENT) break;
        const name = text(key, MAX_PARAMETER_LENGTH);
        if (!/^[a-z][a-z0-9_]*$/.test(name)) continue;

        const value = raw[key];
        if (typeof value === "number" && isFinite(value)) {
            result[name] = Math.round(value * 1000) / 1000;
        } else {
            const label = text(value, MAX_PARAMETER_LENGTH);
            if (!/^[A-Za-z0-9_.-]*$/.test(label)) continue;
            result[name] = label;
        }
        kept++;
    }
    return result;
}

// ⚠️⚠️ THE BATCH ARRIVES AS A JSON STRING, NOT AS AN ARRAY, AND THAT IS A MEASURED CONSTRAINT
// RATHER THAN A STYLE. `docs/TODO.md` § 90.5: this script first declared `events: "JSON"` and the
// service dropped the ENTIRE parameter block, `params: []`, so `action` went missing too and
// every call landed on the default branch again. `String` is the only type these three scripts
// have proven, and `match-record` already passes its whole record the same way.
function parseBatch(raw) {
    if (Array.isArray(raw)) return raw;
    try {
        const parsed = JSON.parse(String(raw || "[]"));
        return Array.isArray(parsed) ? parsed : [];
    } catch (e) {
        return [];
    }
}

function readJson(item, fallback) {
    if (!item || !item.value) return fallback;
    try {
        return typeof item.value === "string" ? JSON.parse(item.value) : item.value;
    } catch (e) {
        return fallback;
    }
}

function emptyProfile() {
    return { Funnel: {}, Counters: {}, Params: {}, Sessions: 0, Batches: 0, LastUtc: "" };
}

function emptyRollup() {
    return { Funnel: {}, Counters: {}, Params: {}, Players: 0, Batches: 0, UpdatedUtc: "" };
}

function addCount(map, name, count) {
    map[name] = (Number(map[name]) || 0) + count;
}

/**
 * Folds a parameter into the aggregate as one counter per distinct label, and as a running
 * mean per numeric key.
 *
 * ⚠️ A NUMBER IS KEPT AS A SUM AND A COUNT, NEVER AS A STORED AVERAGE. `docs/TODO.md` § 89
 * makes the same argument about the career page: a stored rate cannot be re-derived after a
 * change and cannot be added to another period's. Two counts divided at read time can.
 */
function foldParameters(into, eventName, params) {
    for (const key of Object.keys(params)) {
        const value = params[key];
        if (typeof value === "number") {
            const bucket = eventName + "." + key;
            into[bucket] = into[bucket] || { Sum: 0, Count: 0 };
            into[bucket].Sum = (Number(into[bucket].Sum) || 0) + value;
            into[bucket].Count = (Number(into[bucket].Count) || 0) + 1;
        } else if (value) {
            const bucket = eventName + "." + key + "." + value;
            into[bucket] = into[bucket] || { Sum: 0, Count: 0 };
            into[bucket].Count = (Number(into[bucket].Count) || 0) + 1;
        }
    }
}

async function loadRollup(api, projectId) {
    if (typeof api.getCustomItems !== "function" || typeof api.setCustomItem !== "function")
        throw new Error("custom data is not available in this Cloud Save module");

    const response = await api.getCustomItems(projectId, ROLLUP_ID, [ROLLUP_KEY]);
    const rollup = readJson(response.data.results.find(x => x.key === ROLLUP_KEY), emptyRollup());
    rollup.Funnel = rollup.Funnel || {};
    rollup.Counters = rollup.Counters || {};
    rollup.Params = rollup.Params || {};
    return rollup;
}

module.exports = async ({ params, context, logger }) => {
    const api = new DataApi(context);
    const { projectId, playerId } = context;
    const action = String(params.action || "submit");

    const stored = await api.getProtectedItems(projectId, playerId, [KEY]);
    const profile = readJson(stored.data.results.find(x => x.key === KEY), emptyProfile());
    profile.Funnel = profile.Funnel || {};
    profile.Counters = profile.Counters || {};
    profile.Params = profile.Params || {};

    if (action === "report") {
        let rollup = null;
        try {
            rollup = await loadRollup(api, projectId);
        } catch (e) {
            // A rollup this deployment cannot reach is not a failed report. The per-player
            // document is exact either way; the rollup is the convenience view.
            logger.info("rollup unavailable: " + e.message);
        }
        return { profile, rollup, funnel: FUNNEL };
    }

    if (action !== "submit") throw new Error("unknown telemetry action");

    // ⚠️ THE BATCH IS ONE CALL FOR A WHOLE SESSION, WHICH IS `FUTURE.md` § 0.3's ONLY HARD RULE
    // ABOUT CLOUD CODE: *"Call it once per match, never per event."* Telemetry is the feature most
    // able to break that, because every interesting thing in a match is an event. The client
    // counts locally and sends totals; nothing here is shaped to accept a stream.
    const batch = parseBatch(params.events).slice(0, MAX_EVENTS_PER_BATCH);
    const now = new Date().toISOString();

    const firstReached = [];
    let accepted = 0;
    let refused = 0;

    for (const raw of batch) {
        const name = text(raw && raw.Name, MAX_PARAMETER_LENGTH);
        if (EVENTS.indexOf(name) < 0) { refused++; continue; }

        const count = Math.max(1, Math.min(100000, Math.trunc(Number(raw.Count) || 1)));
        const args = columnsFrom(raw.Params);

        addCount(profile.Counters, name, count);
        foldParameters(profile.Params, name, args);
        accepted++;

        // ⚠️⚠️ A FUNNEL STEP IS RECORDED ONCE PER PLAYER, EVER, AND THE FIRST TIMESTAMP WINS.
        // A funnel that could be re-entered is not a funnel, it is a counter with a misleading
        // name: reinstalling, replaying or a client that resends its buffer would each add a
        // second "first" and the conversion rate between two steps could exceed 100 per cent.
        if (FUNNEL.indexOf(name) >= 0 && !profile.Funnel[name]) {
            profile.Funnel[name] = now;
            firstReached.push(name);
        }
    }

    if (accepted > 0) {
        profile.Batches = (Number(profile.Batches) || 0) + 1;
        profile.Sessions = Number(profile.Counters.session_start) || profile.Sessions || 0;
        profile.LastUtc = now;
        await api.setProtectedItem(projectId, playerId, { key: KEY, value: JSON.stringify(profile) });
    }

    // ⚠️⚠️ THE PROJECT-WIDE ROLLUP IS APPROXIMATE AND SAYING SO IS THE POINT. Cloud Save has no
    // atomic increment, so two sessions submitting in the same instant can lose one update. At
    // four to eight concurrent players that is a rounding error against numbers whose job is to
    // say "most people stop here"; at a size where it would matter, this is the wrong storage and
    // the fix is a real analytics sink rather than a lock. The PER-PLAYER document above is exact
    // and is what any number that has to be defended should be recomputed from.
    let rolled = false;
    if (accepted > 0) {
        try {
            const rollup = await loadRollup(api, projectId);
            for (const raw of batch) {
                const name = text(raw && raw.Name, MAX_PARAMETER_LENGTH);
                if (EVENTS.indexOf(name) < 0) continue;
                const count = Math.max(1, Math.min(100000, Math.trunc(Number(raw.Count) || 1)));
                addCount(rollup.Counters, name, count);
                foldParameters(rollup.Params, name, columnsFrom(raw.Params));
            }
            for (const step of firstReached) addCount(rollup.Funnel, step, 1);
            if (firstReached.indexOf(FUNNEL[0]) >= 0) rollup.Players = (Number(rollup.Players) || 0) + 1;
            rollup.Batches = (Number(rollup.Batches) || 0) + 1;
            rollup.UpdatedUtc = now;

            await api.setCustomItem(projectId, ROLLUP_ID,
                { key: ROLLUP_KEY, value: JSON.stringify(rollup) });
            rolled = true;
        } catch (e) {
            // ⚠️ A FAILED ROLLUP MUST NOT FAIL THE SUBMIT. The per-player write above has already
            // landed and it is the exact copy; losing the convenience view would otherwise make
            // the client retry a batch it has already delivered.
            logger.info("rollup write skipped: " + e.message);
        }
    }

    return { accepted, refused, rolled, funnel: profile.Funnel };
};

// ⚠️⚠️ EVERY PARAMETER A SCRIPT USES MUST BE DECLARED HERE OR CLOUD CODE STRIPS IT, AND THE
// FAILURE IS SILENT AND LOOKS LIKE A WORKING ENDPOINT. Measured on 2026-08-30: this endpoint
// answered a call carrying an action with the payload of the branch an ABSENT action falls
// through to. Nothing errors, nothing logs, and the answer is well-formed, so a probe that
// only asserts "it answered" passes. `docs/TODO.md` § 90.5 is the entry.
//
// ⚠️ IF A NEW ACTION NEEDS A NEW PARAMETER, IT GOES IN THIS BLOCK IN THE SAME EDIT.
//
// ⚠️⚠️ NOTHING BUT BARE `name: "Type"` LINES MAY GO INSIDE THE BRACES, AND THAT IS MEASURED.
// `docs/TODO.md` § 90.5: this block first carried a ⚠️-marked comment between its two entries and
// the deploy reported `params: []` for the whole script, twice, while the two scripts whose
// braces held only plain entries kept theirs. A dropped block is not a dropped comment: `action`
// went with it and every call fell back to the default branch, which is the silent failure this
// whole entry exists for. Put the reasoning above the block, as here.
//
// ⚠️ `events` IS A `String`, NOT `JSON`, AND CARRIES A SERIALISED ARRAY. `parseBatch` reads it,
// exactly as `match-record` has always carried its record. `String` and `Numeric` are the only
// two types these three scripts have proven against this project.
module.exports.params = {
    action: "String",
    events: "String",
};

File: Packages/com.tumbangpreso.core/Runtime/Telemetry.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace TumbangPreso.Core
{
    /// <summary>
    /// The event names this game will ever send, and the rules about what may travel with them.
    ///
    /// ⚠️⚠️ THESE NAMES ARE A CONTRACT AND RENAMING ONE IS A BROKEN HISTORY. `FUTURE.md` § 19.3
    /// says it in one line: *"a renamed event is a broken history"*. Nothing errors when a name
    /// changes. What happens instead is that a counter silently restarts at zero, the series it
    /// used to belong to stops growing, and the two halves can never be joined again because
    /// nothing recorded that they were the same question. Choose a name once.
    ///
    /// ⚠️⚠️ `Funnel` IS ORDERED AND APPEND-ONLY, AND THE ORDER IS THE MEANING. "How far did this
    /// player get" is an index comparison, so inserting a step in the middle silently rewrites
    /// what every already-stored profile is claiming: a player recorded at step 3 becomes a
    /// player who reached a step that did not exist when they played. Add to the END, or add a
    /// second funnel. This is `FUTURE.md` § 0.5 rule 5, which `Roster.Slippers` records for the
    /// wire, applied to a list that crosses time instead of a network.
    ///
    /// ⚠️ MIRRORED IN `ugs/cloud-code/telemetry.js`, WHICH REFUSES A NAME IT DOES NOT KNOW.
    /// `docs/TODO.md` § 90.3 is the contract in prose and § 89.6 is the standing argument for why
    /// a file written twice is the accepted cost here. `CareerAndCloudCodeTests` gates the two.
    /// </summary>
    public static class TelemetryEvents
    {
        // The first-launch funnel, in order. `FUTURE.md` § 3: "the most valuable number in this
        // document". It answers where a person who has never played this game stops.
        public const string FirstLaunch = "first_launch";
        public const string FirstSignIn = "first_sign_in";
        public const string FirstMenu = "first_menu";
        public const string FirstQueue = "first_queue";
        public const string FirstMatchStarted = "first_match_started";
        public const string FirstMatchFinished = "first_match_finished";

        // Everything else, which is counted every time rather than once.
        public const string SessionStart = "session_start";
        public const string SessionEnd = "session_end";
        public const string MatchStarted = "match_started";
        public const string MatchFinished = "match_finished";
        public const string MatchLeft = "match_left";
        public const string Pick = "pick";
        public const string SettingsSnapshot = "settings_snapshot";
        public const string Disconnect = "disconnect";

        // ⚠️ ADDED 2026-08-30, AT THE END, WHICH IS THE ONLY SAFE PLACE. `All` is compared
        // against `telemetry.js` by content rather than order, but `Funnel` is compared by both
        // and the two lists are read side by side; appending is the habit that keeps them
        // readable together. This one is not a funnel step and must never become one: it is
        // raised once per FINISHED match, and a funnel position is a thing an install passes
        // exactly once, ever. `docs/TODO.md` § 90.3.
        public const string MatchFrameRate = "match_frame_rate";

        public static readonly string[] Funnel =
        {
            FirstLaunch,
            FirstSignIn,
            FirstMenu,
            FirstQueue,
            FirstMatchStarted,
            FirstMatchFinished,
        };

        public static readonly string[] All =
        {
            FirstLaunch, FirstSignIn, FirstMenu, FirstQueue, FirstMatchStarted, FirstMatchFinished,
            SessionStart, SessionEnd, MatchStarted, MatchFinished, MatchLeft,
            Pick, SettingsSnapshot, Disconnect, MatchFrameRate,
        };
    }

    /// <summary>One counted thing, with the short labels and numbers that describe it.</summary>
    public sealed class TelemetryEvent
    {
        public string Name = "";
        public int Count = 1;
        public readonly Dictionary<string, string> Labels = new Dictionary<string, string>();
        public readonly Dictionary<string, double> Numbers = new Dictionary<string, double>();

        /// <summary>
        /// Two events fold together when they are the same question about the same thing.
        ///
        /// ⚠️ THE NUMBERS ARE PART OF THE KEY AND THAT IS DELIBERATE. Folding `match_finished`
        /// with a duration of 340 s into one with a duration of 512 s would produce a count of
        /// two carrying one of the two durations, and the server averages what it is sent. A
        /// distinct duration is a distinct row; only genuinely identical events collapse.
        /// </summary>
        public string Signature()
        {
            var builder = new StringBuilder(Name);
            AppendSorted(builder, Labels, value => value);
            AppendSorted(builder, Numbers, value => value.ToString("R", System.Globalization.CultureInfo.InvariantCulture));
            return builder.ToString();
        }

        private static void AppendSorted<T>(StringBuilder builder, Dictionary<string, T> map,
                                            Func<T, string> render)
        {
            if (map.Count == 0) return;
            var keys = new List<string>(map.Keys);
            keys.Sort(StringComparer.Ordinal);
            foreach (string key in keys)
            {
                builder.Append('|').Append(key).Append('=').Append(render(map[key]));
            }
        }
    }

    /// <summary>
    /// What may be sent, how much of it, and how far through the funnel somebody has got.
    ///
    /// ⚠️⚠️ NO PERSONALLY IDENTIFYING FIELD MAY EVER REACH AN EVENT, and this is the half of that
    /// rule that runs on the machine the player owns. `ugs/cloud-code/telemetry.js` enforces the
    /// same rule again, because a client is the half somebody can edit. Both halves refuse the
    /// same shapes: a parameter is a bucket label like `hero_strike` or a number, never free text
    /// somebody typed, and no event carries an identifier at all. The server already knows who
    /// called it from the authenticated session, so there is nothing here to get wrong.
    /// </summary>
    public static class TelemetryRules
    {
        // Mirrored in `telemetry.js`. `CareerAndCloudCodeTests` fails if these split.
        public const int MaxEventsPerBatch = 64;
        public const int MaxParametersPerEvent = 8;
        public const int MaxParameterLength = 32;

        /// <summary>
        /// ⚠️⚠️ A FIELD NAME THAT MEANS A PERSON IS REFUSED WHATEVER IT HOLDS. This list is
        /// deliberately about NAMES rather than values, because a value cannot be inspected for
        /// whether it identifies somebody and a name can. It is the cheap check that stops the
        /// expensive mistake: the first person to add `player_name` to an event for debugging
        /// would otherwise ship it, and telemetry is the one system where a mistake is retained
        /// rather than transient.
        /// </summary>
        private static readonly string[] RefusedParameterFragments =
        {
            "name", "email", "player", "account", "handle", "token",
            "address", "serial", "path", "device", "machine", "profile", "session",
        };

        /// <summary>
        /// ⚠️⚠️ THE SHORT ONES ARE MATCHED WHOLE OR AS A SUFFIX, NEVER AS A SUBSTRING, AND THAT
        /// SPLIT IS A MEASURED BUG RATHER THAN A REFINEMENT. `ip` as a fragment refuses
        /// **`slipper`**, which is the tsinelas pick rate: `FUTURE.md` § 3 asks for it by name,
        /// the column would have been stripped on every event, and nothing anywhere would have
        /// said so. `id` as a fragment is the same trap one letter shorter. A two-letter
        /// substring rule is a rule that refuses words it has never heard of.
        /// </summary>
        private static readonly string[] RefusedParameterWords =
        {
            "id", "ip", "mac", "uid", "guid", "user", "uuid",
        };

        public static bool IsKnownEvent(string name)
        {
            if (string.IsNullOrEmpty(name)) return false;
            foreach (string known in TelemetryEvents.All)
                if (known == name) return true;
            return false;
        }

        /// <summary>Position in the funnel, or -1 for an event that is not a funnel step.</summary>
        public static int FunnelIndex(string name)
        {
            for (int i = 0; i < TelemetryEvents.Funnel.Length; i++)
                if (TelemetryEvents.Funnel[i] == name) return i;
            return -1;
        }

        /// <summary>
        /// How far this install has got, given where it had got to and what just happened.
        ///
        /// ⚠️ IT ONLY EVER GOES FORWARD. A funnel that can go backwards is not a funnel: a player
        /// who reaches the menu after finishing a match would otherwise be recorded as having
        /// stopped at the menu, and every conversion rate below that step would be wrong in the
        /// flattering direction.
        /// </summary>
        public static int FurthestFunnelStep(int reached, string name)
        {
            int index = FunnelIndex(name);
            return index > reached ? index : reached;
        }

        /// <summary>Whether a funnel step is new for an install that has reached <paramref name="reached"/>.</summary>
        public static bool IsNewFunnelStep(int reached, string name)
        {
            int index = FunnelIndex(name);
            return index >= 0 && index > reached;
        }

        public static bool IsSafeParameterName(string key)
        {
            if (string.IsNullOrEmpty(key) || key.Length > MaxParameterLength) return false;
            if (key[0] < 'a' || key[0] > 'z') return false;

            foreach (char c in key)
            {
                bool ok = (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '_';
                if (!ok) return false;
            }

            foreach (string fragment in RefusedParameterFragments)
                if (key.Contains(fragment)) return false;

            foreach (string word in RefusedParameterWords)
                if (key == word || key.EndsWith("_" + word) || key.StartsWith(word + "_"))
                    return false;

            return true;
        }

        /// <summary>
        /// A short, boring bucket label, or empty for anything that is not one.
        ///
        /// ⚠️ IT REFUSES RATHER THAN TRUNCATES. `AccountRules.OneLine` truncates a display name
        /// because a clipped name is still that person's name; a clipped telemetry label is a new
        /// value that will never join to the one it came from, which is the same broken history
        /// a renamed event produces.
        /// </summary>
        public static string Label(string raw)
        {
            if (string.IsNullOrEmpty(raw) || raw.Length > MaxParameterLength) return "";
            foreach (char c in raw)
            {
                bool ok = (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') ||
                          (c >= '0' && c <= '9') || c == '_' || c == '.' || c == '-';
                if (!ok) return "";
            }
            return raw;
        }

        /// <summary>
        /// Strips everything the contract does not allow, and answers whether anything is left
        /// worth sending. Refusing the whole event is the right answer for an unknown name; a
        /// refused PARAMETER only costs a column.
        /// </summary>
        public static bool Accept(TelemetryEvent candidate)
        {
            if (candidate == null || !IsKnownEvent(candidate.Name)) return false;
            if (candidate.Count < 1) candidate.Count = 1;

            Prune(candidate.Labels, key => IsSafeParameterName(key));
            Prune(candidate.Numbers, key => IsSafeParameterName(key));

            foreach (var pair in new List<KeyValuePair<string, string>>(candidate.Labels))
                if (Label(pair.Value).Length == 0) candidate.Labels.Remove(pair.Key);

            foreach (var pair in new List<KeyValuePair<string, double>>(candidate.Numbers))
                if (double.IsNaN(pair.Value) || double.IsInfinity(pair.Value))
                    candidate.Numbers.Remove(pair.Key);

            // ⚠️ THE PARAMETER CAP DROPS THE EXTRAS RATHER THAN THE EVENT. An event with nine
            // columns is a design mistake somebody should fix; an event silently not sent is a
            // hole in a history nobody notices for a month.
            TrimTo(candidate.Labels, MaxParametersPerEvent);
            TrimTo(candidate.Numbers, MaxParametersPerEvent - candidate.Labels.Count);

            return true;
        }

        private static void Prune<T>(Dictionary<string, T> map, Func<string, bool> keep)
        {
            foreach (var pair in new List<KeyValuePair<string, T>>(map))
                if (!keep(pair.Key)) map.Remove(pair.Key);
        }

        private static void TrimTo<T>(Dictionary<string, T> map, int limit)
        {
            if (limit < 0) limit = 0;
            if (map.Count <= limit) return;

            var keys = new List<string>(map.Keys);
            keys.Sort(StringComparer.Ordinal);
            for (int i = limit; i < keys.Count; i++) map.Remove(keys[i]);
        }

        /// <summary>
        /// Folds an event into a session's buffer, merging identical ones into a count.
        ///
        /// ⚠️⚠️ THIS IS THE WHOLE REASON TELEMETRY DOES NOT COST ONE CALL PER EVENT. `FUTURE.md`
        /// § 0.3's rule about Cloud Code is *"call it once per match, never per event"*, and
        /// telemetry is the feature most able to break it: a Hero Strike match carrying nine
        /// hundred passive-defence ticks is nine hundred things somebody could send. The buffer
        /// counts locally and one batch goes out per session.
        ///
        /// ⚠️ AND THE BUFFER IS BOUNDED. A session left running overnight in a lobby must not
        /// grow a list until the process dies; past the cap the counts on events already in the
        /// buffer keep rising and no new SHAPE is added, so the numbers stay right for everything
        /// that was already interesting.
        /// </summary>
        public static bool Fold(Dictionary<string, TelemetryEvent> buffer, TelemetryEvent candidate)
        {
            if (buffer == null || !Accept(candidate)) return false;

            string signature = candidate.Signature();
            if (buffer.TryGetValue(signature, out TelemetryEvent existing))
            {
                existing.Count += candidate.Count;
                return true;
            }

            if (buffer.Count >= MaxEventsPerBatch) return false;
            buffer[signature] = candidate;
            return true;
        }
    }
}
